### PR TITLE
Un-hardcode team colors

### PR DESCRIPTION
**Pull request by Gael-de-Sailly**
At the beginning, `ctf_pvp_engine` was designed to handle 12 colors for teams, but most of the further code was then hardcoded to `red` and `blue`, leaving this feature unusable.
Here I've un-hardcoded the colors every time they occur in the code, made the distinction between team name and color (makes possible to use fancy names for teams, independantly of the team's color), and also taken the opportunity to clean up some parts of the code (like match transition or team chest).
It needs substantial changes in `ctf_pvp_engine` as well, see https://github.com/MT-CTF/ctf_pvp_engine/pull/29. All changes are described below.

Example with Purple vs Cyan:
screenshot_20181224_133041

Main changes:
- Colored textures and nodes added:
  - Team chests
  - Doors
  - Hotbars
- Colors API:
  - Created the script `util.lua` that provides 3 functions (they replace `ctf_colors.get_color` and now take names instead of tables)
    - `ctf_colors.get_player_color(player_name)`
    - `ctf_colors.get_team_color(team_name)`
    - `ctf_colors.get_irc_color(player_name)`
  - Completed the list of IRC colors
- Team/Player API:
  - Added `ctf.team_list`, list of team names
- Team base:
  - `ctf_team_base.place` takes the name of the team as its 3rd argument
  - Use node metadata to store the team name in team chests
  - Re-organized the code a bit (use local functions for the most common actions)
- Match transition:
  - Don't remove the loser team when the flag has been captured. This was made to "detect" the winner team by elimination, but prevented to make some useful computations (in `ctf_stats` especially). There are other ways to detect the winner, I'm using the flag clicker's name.
  - Execute `ctf_match.registered_on_winner` in a `minetest.after`, to ensure that all the statistic computation on capture by `ctf_stats` is done before showing the match summary
  - Add callback `ctf_match.register_on_create_teams` that is executed just after creating teams (while `ctf.register_on_new_match` is done before)
- Stats:
  - Removed general team's stats (blue's wins and red's wins on the long run) that have become meaningless since multiple maps have been added. (Maybe to be re-add in another form, per-map?)
  - Removed per-player wins count on red and blue teams for the same reason
  - Detecting the capturer's name is now done with a regular `ctf_match.register_on_capture`

This PR needs a quite extensive testing. I did test it a bit on a local server, until I fixed all bugs I found, but there may be more tests to carry out.

### CODE REVIEW

**Comment by Gael-de-Sailly** [comment]
Travis seems to block because I've modified the submodule to a version that is not merged yet.

**Comment by ClobberXD** [comment]
Since there's a separate PR for `ctf_pvp_engine`, you should revert all commits made to the submodule in this branch to fix Travis.

File: mods/ctf_match/matches.lua [hunk 1]
```diff
@@ -20,6 +20,14 @@ function ctf_match.register_on_new_match(func)
 	table.insert(ctf_match.registered_on_new_match, func)
 end
 
+ctf_match.registered_on_create_teams = {}
+function ctf_match.register_on_create_teams(func)
```
Comment (ClobberXD) [inline mods/ctf_match/matches.lua]: You might want to make a separate PR for this, as this PR is just for un-hardcoding team colours. It makes reviewing and testing easier. :)

File: mods/ctf_match/matches.lua [hunk 2]
```diff
@@ -83,16 +88,12 @@ function ctf_match.create_teams()
...
+	local winner_team = ctf.player(attname).team
 	local fl_team = ctf.team(flag.team)
 	if fl_team and #fl_team.flags == 0 then
 		ctf.action("match", flag.team .. " was defeated.")
-		ctf.remove_team(flag.team)
 		minetest.chat_send_all(flag.team .. " has been defeated!")
 	end
 
-	ctf_match.check_for_winner()
+	ctf_match.win_game(winner_team)
```
Comment (ClobberXD) [inline mods/ctf_match/matches.lua]: You've gotten rid of the extra step of checking for a winner, which was unnecessary as there are only two teams. Very nice. :+1:

But again, this change is completely unrelated to this PR, so this should be a separate PR (maybe combine this with `ctf_match.register_on_create_teams`?)

File: mods/ctf_stats/init.lua [hunk 3]
```diff
@@ -98,7 +77,13 @@ ctf.register_on_save(function()
 end)
 
 function ctf_stats.player_or_nil(name)
-	return ctf_stats.players[name], ctf_stats.current.red[name] or ctf_stats.current.blue[name]
+	local playerstats = ctf_stats.players[name]
```
Comment (ClobberXD) [inline mods/ctf_stats/init.lua]: Could be renamed to `pstats`, for consistency with the rest of the occurences in the files.

File: mods/doors/init.lua [hunk 4]
```diff
@@ -328,7 +331,8 @@ function doors.register(name, def)
 			local dname = name
 			-- If steel doors are placed, append tname to place coloured team-doors instead
 			if name == "doors:door_steel" then
-				dname = name .. "_" .. tname	-- e.g. "doors:door_steel_red"
+				local color = ctf_colors.get_team_color(tname)
+				dname = name .. "_" .. color	-- e.g. "doors:door_steel_red"
```
Comment (ClobberXD) [inline mods/doors/init.lua]:
```suggestion
				dname = name .. "_" .. ctf_colors.get_team_color(tname)
```

Creating a local variable is probably unnecessary, as it's being used only once.

File: mods/ctf_match/matches.lua (hunk 2 above)
Comment (Gael-de-Sailly) [reply mods/ctf_match/matches.lua]: This comes together with un-hardcoding team colors, because we now need the team list at the end of the game (for `ctf_stats`) in order to know the colors. If the list isn't complete, detecting the colors is tricky and hacky.

**Comment by Gael-de-Sailly** [comment]
Many things in this PR doesn't seem related to the initial project at first sight, but they are making it way easier to implement. I could have only un-hardcoded the colors without touching the code much (which would have been a bit hacky), and then remove my hacky code and propose these cleanups in separate PRs, this would represent an extra amount of work, that IMO, is not needed. That's why I prefer to do all of this in one PR, even if it gets off the beaten track.

**Comment by ClobberXD** [comment]
Fair point. If it actually helps with un-hardcoding, then it definitely is relevant to this PR in every way. :)

**Comment by Gael-de-Sailly** [comment]
I see. Will fix it soon.

File: mods/ctf_map/schem_map.lua [hunk 5]
```diff
@@ -160,6 +160,9 @@ function ctf_match.load_map_meta(idx, name)
 	while meta:get("team." .. i) do
 		local tname  = meta:get("team." .. i)
 		local tcolor = meta:get("team." .. i .. ".color")
+		if minetest.global_exists("ctf_colors") then
```
Comment (ClobberXD) [inline mods/ctf_map/schem_map.lua]: Since both `ctf_colors` and `ctf_map` will be present all the time, you could add `ctf_colors` to `depends.txt`, and remove the `if` here.

**Comment by Gael-de-Sailly** [comment]
Rebased.

**Comment by bewlay** [comment]
Tested, everything color-related works: Flag colors, chest colors, base label colors, shirt colors, char colors, kill log colors, summary colors.

Invalid colors cause a crash and a nice error message.

However, when I was testing, the kill log wasn't cleared when finishing a map (only after the first kill on the new map). Switching back to master resolved that, so it must have something to do with this.

**Comment by ClobberXD** [comment]
> However, when I was testing, the kill log wasn't cleared when finishing a map (only after the first kill on the new map). Switching back to master resolved that, so it must have something to do with this.

That's because that was fixed in master only after this PR was created. Nothing to do with the PR itself.

File: mods/ctf/ctf_map/barrier.lua [hunk 6]
```diff
@@ -251,8 +253,8 @@ if minetest.get_modpath("ctf") then
 			local pos = player:get_pos()
 			local privs = minetest.get_player_privs(name)
 			if tname and not privs.fly and privs.interact then
-				if (tname == "blue" and pos.z >= 0) or
-						(tname == "red" and pos.z <= 0) then
+				local base = ctf_map.map.teams[tname].pos
+				if base and base.z * pos.z <= 0 then -- If Z coordinate of spawn and pos have opposite signs, product is negative
```
Comment (ClobberXD) [inline mods/ctf/ctf_map/barrier.lua]: local base = ctf_map.map.teams[tname].pos

				-- If Z coordinate of spawn and pos have opposite signs, product is negative
				if base and base.z * pos.z <= 0 then

File: mods/ctf/ctf_match/matches.lua [hunk 7]
```diff
@@ -51,24 +63,17 @@ end
...
 		ctf.action("match", winner .. " won!")
 		minetest.chat_send_all("Team " .. winner .. " won!")
-		for i = 1, #ctf_match.registered_on_winner do
-			ctf_match.registered_on_winner[i](winner)
-		end
+
+		-- Execute this at the end
 		minetest.after(2, function()
+			for i = 1, #ctf_match.registered_on_winner do
+				ctf_match.registered_on_winner[i](winner)
```
Comment (ClobberXD) [inline mods/ctf/ctf_match/matches.lua]: Why do these callbacks have to execute after 2 seconds?

**Comment by Gael-de-Sailly** [comment]
> However, when I was testing, the kill log wasn't cleared when finishing a map (only after the first kill on the new map). Switching back to master resolved that, so it must have something to do with this.

> That's because that was fixed in master only after this PR was created. Nothing to do with the PR itself.

It's rebased on the most recent master branch, so it can be due to the PR, unless you were testing my old version.

**Comment by bewlay** [comment]
Based on the commit hashes I'm up-to-date on both this PR and the companion PR.

**Comment by ClobberXD** [comment]
Closing due to inactivity. It's sad to see so much effort going waste.

**Comment by Gael-de-Sailly** [comment]
It's still possible that I'll revive it one day. I know I'll have to be more reactive, but to be honest I got bored a bit of Minetest this year and lost my motivation. I hope it will come back...

**Comment by ClobberXD** [comment]
> It's still possible that I'll revive it one day.

That's great to know, thanks! There's a ton of merge conflicts, so it'd be quite a nightmare to get this PR back on track. But starting afresh is actually a better idea, because ctf_pvp_engine is now merged into this repo, and so you could redo this PR with lesser effort.

> I got bored a bit of Minetest this year and lost my motivation. I hope it will come back...

No worries, everyone gets burnt out once in a while. :)
